WolvenKit.App/ViewModels/Documents/RedDocumentTabViewModel.cs

```diff
@@ -13,16 +25,75 @@ public abstract class RedDocumentTabViewModel : ReactiveObject
 
         [Reactive] public bool CanClose { get; set; }
 
-        //[Reactive] public RedDocumentViewModel File { get; set; }
+        [Reactive] public RedDocumentViewModel File { get; set; }
 
         public static IRedType CopiedChunk;
 
         public static List<IRedType> CopiedChunks = new();
 
         public RedDocumentTabViewModel()
         {
+            DeleteEmbeddedFileCommand = new DelegateCommand(_ => ExecuteDeleteEmbeddedFile(), _ => CanDeleteEmbeddedFile());
+            RenameEmbeddedFileCommand = new RelayCommand(ExecuteRenameEmbeddedFile, CanRenameEmbeddedFile);
+        }
 
+        public ICommand DeleteEmbeddedFileCommand { get; private set; }
+        private bool CanDeleteEmbeddedFile() => this is RDTDataViewModel data && data.IsEmbeddedFile;
+        private void ExecuteDeleteEmbeddedFile()
+        {
+            if (this is RDTDataViewModel datavm)
+            {
+                for (int i = 0; i < File.Cr2wFile.EmbeddedFiles.Count; i++)
+                {
+                    var file = File.Cr2wFile.EmbeddedFiles[i];
+                    if (file.Content == datavm.GetData())
+                    {
+                        File.Cr2wFile.EmbeddedFiles.Remove(file);
+                        break;
+                    }
+                }
+                for (int i = 0; i < File.TabItemViewModels.Count; i++)
+                {
+                    var vm = File.TabItemViewModels[i];
+                    if (vm == this)
+                    {
+                        File.TabItemViewModels.Remove(this);
+                        File.SetIsDirty(true);
+                        break;
+                    }
+                }
+            }
         }
 
+        public ICommand RenameEmbeddedFileCommand { get; private set; }
+        private bool CanRenameEmbeddedFile() => this is RDTDataViewModel data && data.IsEmbeddedFile;
+        private async void ExecuteRenameEmbeddedFile()
+        {
+            if (this is RDTDataViewModel datavm)
```

Comment:
Why are we casting `this` to another type? Is this class inherited by others?

Reply:
Yeah, RedDocumentTabViewModel is the parent to all the RDT classes - these need to be defined for all of them due to xaml bindings. This was easier than defining them virtually & overriding 🤷

Reply:
What about a protected function which takes a parameter for `typeof(datavm.GetData())` and a ref for `FilePath`?